pkg/operator/controller/canary/controller.go

```diff
@@ -306,17 +311,45 @@ func (r *reconciler) startCanaryRoutePolling(stop <-chan struct{}) error {
 	return nil
 }
 
-func (r *reconciler) setCanaryFailingStatusCondition() error {
+func (r *reconciler) setCanaryFailingStatusCondition(errors []error) error {
+	errorStrings := deduplicateErrorStrings(errors)
+	if len(errorStrings) > 3 {
+		errorStrings = errorStrings[len(errorStrings)-3:]
+	}
 	cond := operatorv1.OperatorCondition{
 		Type:    ingresscontroller.IngressControllerCanaryCheckSuccessConditionType,
 		Status:  operatorv1.ConditionFalse,
 		Reason:  "CanaryChecksRepetitiveFailures",
-		Message: "Canary route checks for the default ingress controller are failing",
+		Message: fmt.Sprintf("Canary route checks for the default ingress controller are failing. Last %d error messages:\n%s", len(errorStrings), strings.Join(errorStrings, "\n")),
 	}
 
 	return r.setCanaryStatusCondition(cond)
 }
 
+func deduplicateErrorStrings(errors []error) []string {
```

Comment:
is this worth a unit test?

Reply:
I'm not sure; this is a private function at the moment, and I know we were talking about the philosophy of "only test public functions" recently (although I don't recall if we came to any conclusion on whether to follow that or not)

I do think this may be useful enough to try to move it to somewhere like library-go if we wanted to; When I started on this I thought there might already be a deduplication function somewhere like that I could leverage, but I didn't find anything.

If we did move it or make it public, I would definitely want a unit test, though.

Reply:
Yea I agree, I truthfully don't know what I'm doing when it comes to unit testing :man_shrugging: 

The problem is the `Reconcile` function isn't being tested here either, and it may be a  difficult one to ever right a unit test for since it attempts to curl a URL. Though I guess you could stand up a GO echo server, hack the canary route to point to this localhost GO Server, and simulate failures.

That'd be worth doing, but that's a backlog thing. Not going to hold up this PR over that.

Reply:
We have a lot of cases when only `Reconcile` function is public in a package. I agree that the reconcile functions are indeed often not very approachable. Also they have plenty of the test variations which makes the unit testing of `Reconcile` even more difficult.
If we follow strictly the "only test public functions" philosophy, I think that we risk to either stay without the unit tests due to the unnecessary complexity of Reconcile testing for smaller changes or with the minimal coverage.

The way I see it: if the test of the public function is too complex for a smaller change or if it doesn't exist at all - better to have some testing than no testing at all.

Reply:
I think that's fair. I'll add a test for this